Classes/Service/DemandService.php

```diff
@@ -64,7 +70,45 @@ public function getRestrictionsFromDemandArray(
         ExpressionBuilder $expressionBuilder
     ): CompositeExpression
     {
-        // TODO: Most of the code for getting restrictions in this class goes here
+        $demandArray = $this->restrictionsToInt($demandArray);
+        $expressions = [];
+
+        foreach ($demandArray as $key => $restrictions) {
+            if ($restrictions['operator']){
+                $fieldProps = $this->getFieldPropsFromAlias($key);
+            }else{
+                $fieldProps = $this->getFieldPropsFromAlias($restrictions, $key);
+            }
+
+            if (is_string($fieldProps)){
+                $expressions[] = $this->getExpressionFromRestrictions($fieldProps, $restrictions, $expressionBuilder);
+            }else{
+                if ($fieldProps['or']){
+                    $expressionsArr = [];
+
+                    foreach ($fieldProps['or'] as $fieldKey => $field){
+                        foreach ($field as $property) {
+                            $expressionsArr[] = $this->getExpressionFromRestrictions($fieldKey, $property, $expressionBuilder);
+                        }
+                    }
+
+                    $expressions[] = $expressionBuilder->orX(...$expressionsArr);
+                }else{
+                    $expressionsArr = [];
+
+                    foreach ($fieldProps['and'] as $field){
+                        foreach ($field as $fieldKey => $property) {
+                            $expressionsArr[] = $this->getExpressionFromRestrictions($fieldKey, $property, $expressionBuilder);
+                        }
+                    }
+
+                    $expressions[] = $expressionBuilder->andX(...$expressionsArr);
+                }
+            }
+
+        }
+
+        return $expressionBuilder->andX(...$expressions);
```

Comment:
I suggest moving a lot of or all of this into `DemandArrayUtility::toExpression()`. At least as long as it can be run as a static method.